### PR TITLE
Loki: Adds an `interval` paramater to query_range queries allowing a sampling of events to be returned based on the provided interval

### LINKED ISSUE #1779: Interval parameter for log queries
Recently I created #1773 and after merging realized that Grafana defaults to sending a `step` with log queries.  I would consider this a bug but be that as it may keeping #1773 would result in some pretty nasty unexpected behavior when using Grafana against Loki (even if we fixed Grafana old versions would still be sending this parameter)

#1773 was reverted by #1777 however I would still like to add this feature and I can think of a couple ways to do it:

* Use a new parameter, I am suggesting `log_step`
* Use a new header such as: `Accept: application/json;version=2`

Some more thoughts:

I think the `step` functionality is worth having in Loki.  While Loki is a logs database, more specifically it is a string database extended from a metric database (Cortex) and I think it's a valid use case to store string style metrics in Loki (e.g. GPS coordinates, or a more complex json doc of metrics), as such the `step` parameter is of great value at query time to allow range queries which return sane result sizes over different time periods.

At this point it's much too dangerous to just start using the "step" for log style queries, we use it correctly for metric style queries but previously it was just ignored by Loki for log queries.

I think of the two approaches mentioned above, I prefer the new parameter `log_step` for a few reasons, first it makes URL's easy to write and send (setting a header is tedious), second within our code there were also assumptions made about the use of step only being for metric queries and as a result it has specific rules applied to it which don't apply to log queries.  #1773 handled this by parsing the query very early to determine the type before applying these rules to the incoming step parameter.  Using a new parameter would make it more obvious which rules are applied and why within the Loki codebase.

Looking for thoughts/feedback.

### PR DESCRIPTION
**Pull request by slim-bean**
**What this PR does / why we need it**:

This is a possible solution #1779 which creates a new query parameter for range_queries called interval.  It is being marked **experimental** in the docs as it's very possible we will remove this flag one day in favor of a LogQL based solution.

Signed-off-by: Ed Welch <edward.welch@grafana.com>

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: cmd/logcli/main.go [hunk 1]
```diff
@@ -229,7 +229,8 @@ func newQuery(instant bool, cmd *kingpin.CmdClause) *query.Query {
 		cmd.Flag("since", "Lookback window.").Default("1h").DurationVar(&since)
 		cmd.Flag("from", "Start looking for logs at this absolute time (inclusive)").StringVar(&from)
 		cmd.Flag("to", "Stop looking for logs at this absolute time (exclusive)").StringVar(&to)
-		cmd.Flag("step", "Query resolution step width").DurationVar(&q.Step)
+		cmd.Flag("step", "Query resolution step width, for metric queries. Evaluate the query at the specified step over the time range.").DurationVar(&q.Step)
+		cmd.Flag("interval", "Query interval, for log queries. Return entries at the specified interval, ignoring those between.").DurationVar(&q.Interval)
```
Comment (rfratto) [inline cmd/logcli/main.go:233]: I don't want to bikeshed here but I'm worried the difference between step and interval isn't totally clear and might be hard to remember. Would `log_step` or `entry_step` be easier here?

Comment (slim-bean) [reply cmd/logcli/main.go:233]: That would be more confusing to me personally.  

I would prefer to work on making the description clearer if we can do so.

Comment (rfratto) [reply cmd/logcli/main.go:233]: I could be wrong, but I'm concerned about having naming conventions that requires you to pull up the docs to differentiate between the two. I'm okay with this as-is, but I think it's something to think about.

File: pkg/loghttp/params.go [hunk 2]
```diff
@@ -76,6 +76,26 @@ func step(r *http.Request, start, end time.Time) (time.Duration, error) {
...
+		if ts > float64(math.MaxInt64) || ts < float64(math.MinInt64) {
+			return 0, errors.Errorf("cannot parse %q to a valid duration. It overflows int64", value)
+		}
+		return time.Duration(ts), nil
+	}
+	// Or parse as a duration
+	if d, err := model.ParseDuration(value); err == nil {
+		return time.Duration(d), nil
+	}
+	return 0, errors.Errorf("cannot parse %q to a valid duration", value)
```
Comment (cyriltovena) [inline pkg/loghttp/params.go:96]: This whole thing could be a function, parseDuration and reused in step and interval.

File: pkg/logql/engine.go [hunk 3]
```diff
@@ -297,19 +299,33 @@ func PopulateMatrixFromScalar(data promql.Scalar, params LiteralParams) promql.M
...
 }
 
-func readStreams(i iter.EntryIterator, size uint32) (Streams, error) {
+func readStreams(i iter.EntryIterator, size uint32, dir logproto.Direction, interval time.Duration) (Streams, error) {
 	streams := map[string]*logproto.Stream{}
 	respSize := uint32(0)
-	for ; respSize < size && i.Next(); respSize++ {
+	// lastEntry should be a really old time so that the first comparison is always true, we use a negative
+	// value here because many unit tests start at time.Unix(0,0)
+	lastEntry := time.Unix(-100, 0)
```
Comment (cyriltovena) [inline pkg/logql/engine.go:307]: this could be an external const.

var minTime =...

File: pkg/logql/engine.go [hunk 4]
```diff
@@ -297,19 +299,33 @@ func PopulateMatrixFromScalar(data promql.Scalar, params LiteralParams) promql.M
...
 		labels, entry := i.Labels(), i.Entry()
-		stream, ok := streams[labels]
-		if !ok {
-			stream = &logproto.Stream{
-				Labels: labels,
+		forwardShouldOutput := dir == logproto.FORWARD &&
+			(i.Entry().Timestamp.Equal(lastEntry.Add(interval)) || i.Entry().Timestamp.After(lastEntry.Add(interval)))
+		backwardShouldOutput := dir == logproto.BACKWARD &&
+			(i.Entry().Timestamp.Equal(lastEntry.Add(-interval)) || i.Entry().Timestamp.Before(lastEntry.Add(-interval)))
+		// If step == 0 output every line.
```
Comment (cyriltovena) [inline pkg/logql/engine.go:314]:
```suggestion
		// If interval == 0 output every line.
```

**Pull request merged**